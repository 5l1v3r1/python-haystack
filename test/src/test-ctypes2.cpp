
#include <stdio.h>
#include <stdlib.h>
#include <iostream>

#ifdef PYTHON_BUILD
#endif

// structs
struct sA {
  int a;
};

struct sB : sA {
  unsigned int b;
};

struct sC : sB {
  private:
  unsigned int c;
};

struct sD : sB {
  public:
  unsigned int d;
};

// classes
class cA {
  int a;
};

class cB : cA {
  unsigned int b;
};

class cC : cB {
  private:
  unsigned int c;
};

class cD : cB {
  public:
  unsigned int d;
};

class cE : cD, cC {
  public:
  unsigned int e;
};


int test_classes()
{
  std::cout << " -- classes --" << std::endl;

  cA * a = new cA();
  cB * b = new cB();
  cC * c = new cC();
  cD * d = new cD();
  cE * e = new cE();
  
  printf(" a is at 0x%x , size: %d \n", (unsigned int) a, sizeof(cA));
  printf(" b is at 0x%x , size: %d \n", (unsigned int) b, sizeof(cB));
  printf(" c is at 0x%x , size: %d \n", (unsigned int) c, sizeof(cC));
  printf(" d is at 0x%x , size: %d \n", (unsigned int) d, sizeof(cD));
  printf(" e is at 0x%x , size: %d \n", (unsigned int) e, sizeof(cE));

  std::cout << " -- end classes --" << std::endl;
  
  return 0;
}

int test_structs()
{
  sA * a = (sA * ) malloc(sizeof(sA));
  sB * b = (sB * ) malloc(sizeof(sB));
  sC * c = (sC * ) malloc(sizeof(sC));
  sD * d = (sD * ) malloc(sizeof(sD));
}

int main(){
  
  test_structs();
  test_classes();
  return 0;
}
