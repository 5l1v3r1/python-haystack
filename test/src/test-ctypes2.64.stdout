START
 -- structs --
 a is at 0x19eac20 size: 4 
 b is at 0x19eac40 size: 8 
 c is at 0x19eac60 size: 12 
 d is at 0x19eac80 size: 12 
 -- end structs --
 -- classes --
 a is at 0x19eaca0 size: 4 
 b is at 0x19eacc0 size: 8 
 c is at 0x19eace0 size: 12 
 d is at 0x19ead00 size: 12 
 e is at 0x19ead20 size: 28 
 -- end classes --
